src/search.rs

```diff
@@ -0,0 +1,146 @@
+use crate::search_error::SearchError;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+pub fn search(directory: &Path) -> Result<PathBuf, SearchError> {
+  let mut files = Vec::new();
+  let dir = fs::read_dir(directory).map_err(|io_error| SearchError::Io {
+    io_error,
+    directory: directory.to_owned(),
+  })?;
+  for entry in dir {
+    let entry = entry.map_err(|io_error| SearchError::Io {
+      io_error,
+      directory: directory.to_owned(),
+    })?;
+    if let Some(name) = entry.file_name().to_str() {
+      use caseless::default_caseless_match_str;
+      if default_caseless_match_str(name, "justfile") {
+        files.push(entry.path());
+      }
+    }
+  }
+  if files.len() == 1 {
+    Ok(files.pop().unwrap())
+  } else if files.len() > 1 {
+    Err(SearchError::MultipleCandidates { candidates: files })
+  } else if let Some(parent_dir) = directory.parent() {
+    search(parent_dir)
+  } else {
+    Err(SearchError::NotFound)
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use crate::search::search;
```

Comment:
You can do `use super::*;` to bring all items in the parent module into scope.

Reply:
Sorry, but I'm not sure I understood this moment correctly. If we want to call `justfile` function with full path name(search::justfile()), then we need to import `search` mod here explicitly: `use crate::search;` Thus, should I do `use super::*;` to import `SearchError` only?

Reply:
Inside of `src/common.rs`, there's a use statement which contains modules that are used directly, you can add `search` to that use statement, like so:

// modules
pub(crate) use crate::{color, doc, hint, mode, search};

Then, if you do `use crate::common::*;` at the top of `src/search.rs`, the `use super::*;` will bring everything into scope, including the `search` module.

Reply:
Thank you for your answer! Fixed it in the last commit.